librepo/downloader.c

```diff
@@ -1660,6 +1676,22 @@ check_transfer_statuses(LrDownload *dd, GError **err)
                 // Truncate file - remove downloaded garbage (error html page etc.)
                 if (!truncate_transfer_file(target, err))
                     return FALSE;
+
+                if ((target->mirror->max_parallel_connections = -1 ||
+                     target->mirror->max_parallel_connections > 1) &&
+                    (target->mirror->running_transfers > 0 ||
+                     (target->mirror->successful_transfers > 0 &&
+                      target->mirror->failed_transfers < dd->max_connection_per_host)))
```

Comment:
I would refactor the conditions to be cleaner. Right now it is hard to understand what it does.